### PR TITLE
Added support for 12 Bit and fixed issue with ESP boards.

### PR DESCRIPTION
**Pull request by erikbdev**
- Fixed issue with some ESP boards returning 0 when the audio was max value. This was caused by analogRead returning 1024 for some esp boards, and shifting to 2 bits to the right would return 256 and casting it to uint8_t would return 0. More info of the issue here: https://github.com/NicoHood/MSGEQ7/issues/14
- change smooth type to uint16 to support 10 and 12 Bit.
- changed casting of smoothing algorithm to uint32 to support 10 and 12 bit smoothing.

Should I implement analogReadResolution in the future? For some systems like AVR Duo or ESP32 who support 12 bit?

### CODE REVIEW

File: src/MSGEQ7.hpp [hunk 1]
```diff
@@ -64,7 +64,7 @@ CMSGEQ7(void)
 // IC setup
 //================================================================================
 
-template <uint8_t smooth, uint8_t resetPin, uint8_t strobePin, uint8_t firstAnalogPin, uint8_t ...analogPins>
+template <uint16_t smooth, uint8_t resetPin, uint8_t strobePin, uint8_t firstAnalogPin, uint8_t ...analogPins>
```
Comment (NicoHood) [inline src/MSGEQ7.hpp]: shouldnt this be MSGEQ7_data_t instead of uint16_t? for all other functions as well.

File: src/MSGEQ7.hpp [hunk 2]
```diff
@@ -116,13 +116,8 @@ read(void){
...
 		frequency f = (frequency){
-#ifdef MSGEQ7_10BIT
-			analogRead(firstAnalogPin),
-				analogRead(analogPins)...
-#else
-			MSGEQ7_data_t(analogRead(firstAnalogPin) >> 2),
-				MSGEQ7_data_t(analogRead(analogPins) >> 2)...
-#endif
+			getAnalogReadPin(firstAnalogPin),
+			getAnalogReadPin(analogPins)...
```
Comment (NicoHood) [inline src/MSGEQ7.hpp:119]: Why did you remove the 8 bit variant?

File: src/MSGEQ7.h [hunk 3]
```diff
@@ -104,6 +111,10 @@ class CMSGEQ7{
 		// little hack to take as many arguments as possible
 		// to execute several functions for the analogPins
 	}
+
+	// This calculates the value based on what MSGEQ7_data_t is defined.
+	// This assumes the analog pin resolution is 10 bit.
```
Comment (NicoHood) [inline src/MSGEQ7.h]: I am not sure if this comment is correct.

File: src/MSGEQ7.hpp [hunk 4]
```diff
@@ -220,3 +215,22 @@ getVolume(void){
...
+template <uint16_t smooth, uint8_t resetPin, uint8_t strobePin, uint8_t firstAnalogPin, uint8_t ...analogPins>
+MSGEQ7_data_t CMSGEQ7<smooth, resetPin, strobePin, firstAnalogPin, analogPins ...>::
+getAnalogReadPin(uint8_t pin) {
+	// This assumes the analogReadResolution is already at 10 bit.
+	int read = analogRead(pin);
+	#if defined(MSGEQ7_10BIT) || defined(MSGEQ7_12BIT)
+		// Return the value as int.
+		return read;
+	#else
+		// get max resolution of analog pin and map it to 8 bit range. Some systems, esp8266, have range from 0 - 1024 so there is a max range of 1023.
```
Comment (NicoHood) [inline src/MSGEQ7.hpp]: Sorry, I am a bit confused. Doesnt the esp has 12 bite analog value? why are you writing about an 10 bit value here? Also the bitshifting depends on the input value. If the input is 10 bit (avr) you need to shift twice, on esp 12 bit you need to shift 4 times. wouldnt it be better to detect the MCU family and shift the bits accordingly?

Comment (erikbdev) [reply src/MSGEQ7.hpp]: > Sorry, I am a bit confused. Doesnt the esp has 12 bite analog value?


Yes. Well, depends on the board. On the ESP8266, which is a variant of ESP, only allows 10 Bit analog value, but on the ESP32, which is another variant of ESP, is able to change the analog value from 8 Bit to 12 Bit, just like the Due, Zero and MKR Family. More info here: https://www.arduino.cc/reference/en/language/functions/zero-due-mkr-family/analogreadresolution/.



> why are you writing about an 10 bit value here?

I wrote the 10 bit value there because the user is able to change the analog values. Let's say I decided to change the analog value from 10 bit to 12 Bit, well then that mean that you will have to shift the bits 4 times, which is what you said, but there is no way to see what analog value is the pin, but I could not find a method that let us know what the analog value is for the current pin.

> wouldnt it be better to detect the MCU family and shift the bits accordingly?

That would be better but Is there a method to detect the current resolution of a bit? Unless if we set the analogReadResolution?

File: src/MSGEQ7.hpp (hunk 1 above)
Comment (erikbdev) [reply src/MSGEQ7.hpp]: Yes, it should.

File: src/MSGEQ7.hpp (hunk 2 above)
Comment (erikbdev) [reply src/MSGEQ7.hpp:119]: I removed the 8 Bit variant because it's already set in the getAnalogReadPin if MSGEQ7_10BIT or MSGEQ7_12BIT is not defined.

File: src/MSGEQ7.h (hunk 3 above)
Comment (erikbdev) [reply src/MSGEQ7.h]: > I am not sure if this comment is correct.

That comment is for the method getAnalogReadPin. Depending on what Bit value is set, it returns the MSGEQ7_data_t value.

File: src/MSGEQ7.hpp [hunk 5]
```diff
@@ -220,3 +214,21 @@ getVolume(void){
 	// return the average of all channels
 	return vol / 7;
 }
+
+template <MSGEQ7_data_t smooth, uint8_t resetPin, uint8_t strobePin, uint8_t firstAnalogPin, uint8_t ...analogPins>
+MSGEQ7_data_t CMSGEQ7<smooth, resetPin, strobePin, firstAnalogPin, analogPins ...>::
+getAnalogReadPin(uint8_t pin) {
+	int read = analogRead(pin);
```
Comment (NicoHood) [inline src/MSGEQ7.hpp:221]: we could use `auto` here, right?

File: src/MSGEQ7.hpp [hunk 6]
```diff
@@ -220,3 +214,21 @@ getVolume(void){
...
+	#if defined(MSGEQ7_10BIT) || defined(MSGEQ7_12BIT)
+		// Return the value as int.
+		return read;
+	#else
+		// Some systems, esp8266, have range from 0 - 1024 so there is a max range of 1023.
+		uint16_t max10BitVal = 1023;
```
Comment (NicoHood) [inline src/MSGEQ7.hpp:227]: I am not sure, does this give compiler warnings, because you compare signed vs unsigned? Maybe a #define should be used here? I am not sure, just want to think about it...

File: src/MSGEQ7.hpp (hunk 4 above)
Comment (NicoHood) [reply src/MSGEQ7.hpp]: You can do arduino board checks like #ifdef ARDUINO_NANO (or similar). We could add that for esp and do a 4x shift? That is hard to explain...

AVR board always have a 10 bit measure resolution. We decide on the output resolution in our code, and we normally want 8 bits. So we need to shift twice `>>2`

On ESP (as far as I understand) the analog value is always measured in 12 bits (correct?). we need to shift twice for 10 bit and 4th for 8 bit then. So we need to shift, depending on the analog input range.

Comment (erikbdev) [reply src/MSGEQ7.hpp]: > On ESP (as far as I understand) the analog value is always measured in 12 bits (correct?). we need to shift twice for 10 bit and 4th for 8 bit then. So we need to shift, depending on the analog input range.

Some ESP boards, like the ESP8266, measures the analog values ONLY at 10 bit. Others, like the ESP32, allows to change the bit value from 8 - 12 bit (and theoretically even more) but it's recommended to stay in the 8-12 bit. Zero, Due, and the MKR family also has these exact features. 

> we need to shift twice for 10 bit and 4th for 8 bit then. So we need to shift, depending on the analog input range.

I'm going to cancel this pull request since this is trying to add support to 12 bit, but I do not think It's ready yet, but what I am trying to fix is the issue with the ESP board value changing to zero and the primary reason for that is for some reason the 10 bit analog value of the ESP8266 ranges from 0 - 1024 rather than 0 - 1023. If you shift 1024 twice, you get 256, but 256 does not fit in the `uint8_t` range, which returns 0.

**Comment by NicoHood** [comment]
Why did you close the PR?

**Comment by erikbdev** [comment]
> Why did you close the PR?

Closed it since I have not worked on it for the longest, but I might come back and see if I can fix it once I get another MSGEQ7.
